// mongoose-validatefilter/libs/tool.js
// 工具类函数


var rEmail = /^[\w!\#\$%&'*+\/=?^`{|}~.\-]+@([\w-]+\.)+(com|net|org|hk|cn|gov|biz|info|cc|tv|mobi|name|asia|tw|sh|ac|io|tm|travel|ws|us|sc|mn|ag|vc|la|bz|in|cm|co|tel|me|pro|pw|im|xn--[a-zA-Z\d]+)$/i;
var rUrl = /^https?:\/\/(\w(\:\w)?@)?([0-9a-z_-]+\.)*?([a-z0-9-]+\.[a-z]{2,6}(\.[a-z]{2})?(\:[0-9]{2,6})?)((\/[^?#<>\/\\*':]*)+(\?[^#]*)?(#.*)?)?$/;

module.exports = {

    /**
     * 判断数据类型
     * @link http://javascript.ruanyifeng.com/stdlib/object.html#toc3
     * @param  {*} Anything
     * @return {String} 数据类型
     * 包括：array/boolean/element/function/infinite/nan/
     *     null/number/object/regexp/string/undefined
     * @version 1.4
     * 2013年6月28日12:42:11
     * 2013年7月1日18:08:02
     * 2013年7月24日23:30:01
     * 2013年12月3日9:45:13
     * 2014年2月11日21:40:55
     */

    type: function(any) {
        var ret = Object.prototype.toString.call(any).match(/\s(.*?)\]/i)[1].toLowerCase();
        if (ret == 'number' && isNaN(any)) return 'nan';
        if (/element/.test(ret)) return 'element';
        return ret;
    },


    /**
     * 判断是否为 email
     * @param  {String}
     * @return {Boolean}
     * @version 1.0
     * 2014年4月18日21:34:18
     */
    isEmail: function(any) {
        return rEmail.test(String(any));
    },


    /**
     * 判断是否为 url
     * @param  {String}
     * @return {Boolean}
     * @version 1.0
     * 2014年4月18日21:34:14
     */
    isUrl: function(any) {
        return rUrl.test(String(any));
    },


    /**
     * 判断是否为空对象
     * undefined 或 null
     * @param  {*}        对象
     * @return {Boolean}
     * @version 1.0
     * 2014年4月20日19:18:17
     */
    isUndefinedOrNull: function(any) {
        return any === undefined || any === null;
    },


    /**
     * 判断是否为字符串或数字
     * @param  {*}        对象
     * @return {Boolean}
     * @version 1.0
     * 2014年4月20日19:21:25
     */
    isStringOrNumber: function(any) {
        return this.type(any) === 'string' || this.type(any) === 'number';
    },


    /**
     * 判断元素在数组中的第几个
     * @param  {*}      值
     * @param  {Array}  数据
     * @return {Number} 如果存在则返回索引值，否则返回-1
     * @version 1.0
     * 2014年2月16日1:32:47
     */

    inArray: function(value, array) {
        for (var i = 0, j = array.length; i < j; i++) {
            if (array[i] === value) return i;
        }
        return -1;
    },


    /**
     * 获得验证方法
     * @param  {Object} rule对象
     * @return {String} rule类型
     * @version 1.0
     * 2014年4月18日21:32:08
     */
    getValidateTypeOfRule: function(rule) {
        for (var i in rule) {
            if (i === 'msg' || i === 'trim' || i === 'exist') {} else {
                return i;
            }
        }
        return '';
    },

};
